File: src/lexer.cpp
#include "lexer.hpp"
#include <stdexcept>
#include <cmath>

namespace riddle
{
    std::vector<std::unique_ptr<const token>> lexer::parse(std::istream &is)
    {
        std::vector<std::unique_ptr<const token>> tokens;

        current_state = START;
        line = 1;
        start = 0;
        text.clear();

        while (is.good())
        {
            switch (auto c = is.get())
            {
            case '/':
                switch (is.peek())
                {
                case '/': // single line comment
                    is.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
                    ++line;
                    start = 0;
                    break;
                case '*': // multi-line comment
                    is.ignore();
                    while (is.good())
                    {
                        if (is.get() == '*' && is.peek() == '/')
                        {
                            is.ignore();
                            break;
                        }
                        if (is.peek() == '\n')
                            ++line;
                    }
                    start = 0;
                    break;
                default:
                    text.push_back(c);
                    current_state = SLASH;
                    tokens.push_back(finish_token());
                }
                break;
            case 'b':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'o'))
                        text.push_back('o');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'o'))
                        text.push_back('o');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'l'))
                        text.push_back('l');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = BOOL;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'c':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'l'))
                        text.push_back('l');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'a'))
                        text.push_back('a');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 's'))
                        text.push_back('s');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 's'))
                        text.push_back('s');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = CLASS;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'e':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'n'))
                        text.push_back('n');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'u'))
                        text.push_back('u');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'm'))
                        text.push_back('m');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = ENUM;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'f':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'a'))
                    {
                        text.push_back('a');
                        if (match(is, 'c'))
                        {
                            text.push_back('c');
                            if (match(is, 't'))
                                text.push_back('t');
                            else
                                break;
                            current_state = FACT;
                            tokens.push_back(finish_token());
                        }
                        else if (match(is, 'l'))
                        {
                            text.push_back('l');
                            if (match(is, 's'))
                                text.push_back('s');
                            else
                                break;
                            if (match(is, 'e'))
                                text.push_back('e');
                            else
                                break;
                            current_state = Bool;
                            tokens.push_back(finish_token());
                        }
                        else
                        {
                            current_state = ID;
                            text.push_back(c);
                        }
                    }
                    else if (match(is, 'o'))
                    {
                        text.push_back('o');
                        if (match(is, 'r'))
                            text.push_back('r');
                        else
                            break;
                        current_state = FOR;
                        tokens.push_back(finish_token());
                    }
                    else
                        current_state = ID;
                }
                else
                    text.push_back(c);
                break;
            case 'g':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'o'))
                        text.push_back('o');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'a'))
                        text.push_back('a');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'l'))
                        text.push_back('l');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = GOAL;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'i':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'n'))
                        text.push_back('n');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 't'))
                        text.push_back('t');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = INT;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'n':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'e'))
                        text.push_back('e');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'w'))
                        text.push_back('w');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = NEW;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'o':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'r'))
                        text.push_back('r');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = OR;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'p':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'r'))
                        text.push_back('r');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'e'))
                        text.push_back('e');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'd'))
                        text.push_back('d');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'i'))
                        text.push_back('i');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'c'))
                        text.push_back('c');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'a'))
                        text.push_back('a');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 't'))
                        text.push_back('t');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'e'))
                        text.push_back('e');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = PREDICATE;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 'r':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'e'))
                        text.push_back('e');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'a'))
                    {
                        text.push_back('a');
                        if (match(is, 'l'))
                            text.push_back('l');
                        else
                            break;
                        current_state = REAL;
                        tokens.push_back(finish_token());
                    }
                    else if (match(is, 't'))
                    {
                        text.push_back('t');
                        if (match(is, 'u'))
                            text.push_back('u');
                        else
                            break;
                        if (match(is, 'r'))
                            text.push_back('r');
                        else
                            break;
                        if (match(is, 'n'))
                            text.push_back('n');
                        else
                            break;
                        current_state = RETURN;
                        tokens.push_back(finish_token());
                    }
                    else
                        current_state = ID;
                }
                else
                    text.push_back(c);
                break;
            case 's':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 't'))
                        text.push_back('t');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'r'))
                        text.push_back('r');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'i'))
                        text.push_back('i');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'n'))
                        text.push_back('n');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'g'))
                        text.push_back('g');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = STRING;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case 't':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'h'))
                    {
                        text.push_back('h');
                        if (match(is, 'i'))
                            text.push_back('i');
                        else
                            break;
                        if (match(is, 's'))
                            text.push_back('s');
                        else
                            break;
                        current_state = THIS;
                        tokens.push_back(finish_token());
                    }
                    else if (match(is, 'i'))
                    {
                        text.push_back('i');
                        if (match(is, 'm'))
                            text.push_back('m');
                        else
                            break;
                        if (match(is, 'e'))
                            text.push_back('e');
                        else
                            break;
                        current_state = TIME;
                        tokens.push_back(finish_token());
                    }
                    else if (match(is, 'r'))
                    {
                        text.push_back('r');
                        if (match(is, 'u'))
                            text.push_back('u');
                        else
                            break;
                        if (match(is, 'e'))
                            text.push_back('e');
                        else
                            break;
                        current_state = Bool;
                        tokens.push_back(finish_token());
                    }
                    else
                        current_state = ID;
                }
                else
                    text.push_back(c);
                break;
            case 'v':
                if (current_state == START)
                {
                    text.push_back(c);
                    if (match(is, 'o'))
                        text.push_back('o');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'i'))
                        text.push_back('i');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    if (match(is, 'd'))
                        text.push_back('d');
                    else
                    {
                        current_state = ID;
                        break;
                    }
                    current_state = VOID;
                    tokens.push_back(finish_token());
                }
                else
                    text.push_back(c);
                break;
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9':
                if (current_state == START)
                    current_state = Int;
                else if (current_state != ID && current_state != Int && current_state != Real)
                    tokens.push_back(finish_token());
                text.push_back(c);
                break;
            case '.':
                if ((current_state == START && std::isdigit(is.peek())) || current_state == Int)
                {
                    current_state = Real;
                    text.push_back(c);
                }
                else
                {
                    if (!text.empty())
                        tokens.push_back(finish_token());
                    current_state = DOT;
                    text.push_back(c);
                    tokens.push_back(finish_token());
                }
                break;
            case '"':
                switch (current_state)
                {
                case START:
                    current_state = String;
                    text.push_back(c);
                    break;
                case String:
                    text.push_back(c);
                    tokens.push_back(finish_token());
                    break;
                default:
                    throw std::runtime_error("Unexpected character: " + std::string(1, c));
                }
                break;
            case '(':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = LPAREN;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case ')':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = RPAREN;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '{':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = LBRACE;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '}':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = RBRACE;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '[':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = LBRACKET;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case ']':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = RBRACKET;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case ',':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = COMMA;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case ':':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = COLON;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case ';':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = SEMICOLON;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '+':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = PLUS;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '-':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = MINUS;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '*':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = STAR;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '&':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = AMP;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '|':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = BAR;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '?':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = QUESTION;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '=':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = EQ;
                text.push_back(c);
                if (is.peek() == '=')
                {
                    is.ignore();
                    current_state = EQEQ;
                    text.push_back('=');
                }
                tokens.push_back(finish_token());
                break;
            case '>':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = GT;
                text.push_back(c);
                if (is.peek() == '=')
                {
                    is.ignore();
                    current_state = GTEQ;
                    text.push_back('=');
                }
                tokens.push_back(finish_token());
                break;
            case '<':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = LT;
                text.push_back(c);
                if (is.peek() == '=')
                {
                    is.ignore();
                    current_state = LTEQ;
                    text.push_back('=');
                }
                tokens.push_back(finish_token());
                break;
            case '!':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = BANG;
                text.push_back(c);
                if (is.peek() == '=')
                {
                    is.ignore();
                    current_state = BANGEQ;
                    text.push_back('=');
                }
                tokens.push_back(finish_token());
                break;
            case '^':
                if (!text.empty())
                    tokens.push_back(finish_token());
                current_state = CARET;
                text.push_back(c);
                tokens.push_back(finish_token());
                break;
            case '\r':
                if (current_state == String)
                    text.push_back(c);
                else
                {
                    if (!text.empty())
                        tokens.push_back(finish_token());
                    start = 0;
                }
                break;
            case '\n':
                if (current_state == String)
                    text.push_back(c);
                else
                {
                    if (!text.empty())
                        tokens.push_back(finish_token());
                    start = 0;
                    ++line;
                }
                break;
            case ' ':
            case '\t':
                if (current_state == String)
                    text.push_back(c);
                else
                {
                    if (!text.empty())
                        tokens.push_back(finish_token());
                    ++start;
                }
                break;
            case EOF:
                if (!text.empty())
                    tokens.push_back(finish_token());
                tokens.push_back(std::make_unique<token>(EoF, line, start, start));
                break;
            default:
                if (current_state == START)
                    current_state = ID;
                text.push_back(c);
            }
        }

        return tokens;
    }

    bool lexer::match(std::istream &is, char expected) noexcept
    {
        if (is.peek() == expected)
        {
            is.ignore();
            return true;
        }
        return false;
    }

    std::unique_ptr<const token> lexer::finish_token() noexcept
    {
        std::unique_ptr<token> tok;
        auto end = start + text.size() - 1;
        switch (current_state)
        {
        case Bool:
            tok = std::make_unique<bool_token>(text == "true", line, start, end);
            break;
        case Int:
            tok = std::make_unique<int_token>(std::stoll(text), line, start, end);
            break;
        case Real:
        {
            size_t dot_pos = text.find('.');
            std::string intgr = text.substr(0, dot_pos);
            std::string dec = text.substr(dot_pos + 1);
            tok = std::make_unique<real_token>(utils::rational(static_cast<INT_TYPE>(std::stol(intgr + dec)), static_cast<INT_TYPE>(std::pow(10, dec.size()))), line, start, end);
            break;
        }
        case String:
            tok = std::make_unique<string_token>(std::move(text.substr(1, text.size() - 2)), line, start, end);
            break;
        case ID:
            tok = std::make_unique<id_token>(std::move(text), line, start, end);
            break;
        default:
            tok = std::make_unique<token>(current_state, line, start, end);
        }
        start = end + 1;
        current_state = START;
        text.clear();
        return tok;
    }
} // namespace riddle

File: src/lexer.rs
use std::iter::Peekable;
use std::str::Chars;

#[derive(Debug, PartialEq, Clone)]
pub(crate) enum Token {
    Identifier(String),
    BoolLiteral(bool),
    IntLiteral(i64),
    RealLiteral(i64, i64),
    StringLiteral(String),
    Plus,
    Minus,
    Asterisk,
    Slash,
    Amp,
    Bar,
    Dot,
    LParen,
    RParen,
    LBracket,
    RBracket,
    LBrace,
    RBrace,
    Comma,
    Colon,
    Equal,
    EqualEqual,
    Not,
    NotEqual,
    LessThan,
    LessEqual,
    GreaterThan,
    GreaterEqual,
    Semicolon,
    Bool,
    Int,
    Real,
    String,
    Class,
    Predicate,
    New,
    For,
    This,
    Void,
    Return,
    Fact,
    Goal,
    Or,
    Eof,
}

pub(crate) struct Lexer<'a> {
    input: Peekable<Chars<'a>>,
}

impl<'a> Lexer<'a> {
    pub(crate) fn new(input: &'a str) -> Self {
        Lexer { input: input.chars().peekable() }
    }

    pub(crate) fn next_token(&mut self) -> Token {
        self.skip_whitespace_and_comments();
        match self.input.peek() {
            Some(&ch) => match ch {
                '+' => {
                    self.input.next();
                    Token::Plus
                }
                '-' => {
                    self.input.next();
                    Token::Minus
                }
                '*' => {
                    self.input.next();
                    Token::Asterisk
                }
                '/' => {
                    self.input.next();
                    Token::Slash
                }
                '&' => {
                    self.input.next();
                    Token::Amp
                }
                '|' => {
                    self.input.next();
                    Token::Bar
                }
                '.' => {
                    let mut lookahead = self.input.clone();
                    lookahead.next();
                    if let Some(&ch) = lookahead.peek() {
                        if ch.is_ascii_digit() {
                            self.read_number()
                        } else {
                            self.input.next();
                            Token::Dot
                        }
                    } else {
                        self.input.next();
                        Token::Dot
                    }
                }
                '(' => {
                    self.input.next();
                    Token::LParen
                }
                ')' => {
                    self.input.next();
                    Token::RParen
                }
                '[' => {
                    self.input.next();
                    Token::LBracket
                }
                ']' => {
                    self.input.next();
                    Token::RBracket
                }
                '{' => {
                    self.input.next();
                    Token::LBrace
                }
                '}' => {
                    self.input.next();
                    Token::RBrace
                }
                ',' => {
                    self.input.next();
                    Token::Comma
                }
                ':' => {
                    self.input.next();
                    Token::Colon
                }
                '=' => {
                    self.input.next();
                    if let Some(&'=') = self.input.peek() {
                        self.input.next();
                        Token::EqualEqual
                    } else {
                        Token::Equal
                    }
                }
                '!' => {
                    self.input.next();
                    if let Some(&'=') = self.input.peek() {
                        self.input.next();
                        Token::NotEqual
                    } else {
                        Token::Not
                    }
                }
                '<' => {
                    self.input.next();
                    if let Some(&'=') = self.input.peek() {
                        self.input.next();
                        Token::LessEqual
                    } else {
                        Token::LessThan
                    }
                }
                '>' => {
                    self.input.next();
                    if let Some(&'=') = self.input.peek() {
                        self.input.next();
                        Token::GreaterEqual
                    } else {
                        Token::GreaterThan
                    }
                }
                ';' => {
                    self.input.next();
                    Token::Semicolon
                }
                '"' => {
                    self.input.next(); // consume opening quote
                    let mut string = String::new();
                    while let Some(&ch) = self.input.peek() {
                        if ch == '"' {
                            self.input.next(); // consume closing quote
                            break;
                        } else {
                            string.push(ch);
                            self.input.next();
                        }
                    }
                    Token::StringLiteral(string)
                }
                '0'..='9' => self.read_number(),
                'a'..='z' | 'A'..='Z' | '_' => self.read_identifier(),
                _ => {
                    self.input.next();
                    self.next_token()
                }
            },
            None => Token::Eof,
        }
    }

    fn skip_whitespace(&mut self) {
        while let Some(&ch) = self.input.peek() {
            if ch.is_whitespace() {
                self.input.next();
            } else {
                break;
            }
        }
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            self.skip_whitespace();

            let mut lookahead = self.input.clone();
            match (lookahead.next(), lookahead.next()) {
                (Some('/'), Some('/')) => {
                    self.input.next();
                    self.input.next();
                    for ch in self.input.by_ref() {
                        if ch == '\n' {
                            break;
                        }
                    }
                }
                (Some('/'), Some('*')) => {
                    self.input.next();
                    self.input.next();
                    let mut prev = '\0';
                    for ch in self.input.by_ref() {
                        if prev == '*' && ch == '/' {
                            break;
                        }
                        prev = ch;
                    }
                }
                _ => break,
            }
        }
    }

    fn read_number(&mut self) -> Token {
        let mut number = String::new();
        let mut has_decimal_point = false;

        while let Some(&ch) = self.input.peek() {
            if ch.is_ascii_digit() {
                number.push(ch);
                self.input.next();
            } else if ch == '.' && !has_decimal_point {
                has_decimal_point = true;
                number.push(ch);
                self.input.next();
            } else {
                break;
            }
        }

        if has_decimal_point {
            let mut parts = number.splitn(2, '.');
            let int_part = parts.next().unwrap_or("0");
            let frac_part = parts.next().unwrap_or("0");
            let int_value = int_part.parse::<i64>().unwrap_or(0);
            let frac_value = frac_part.parse::<i64>().unwrap_or(0);
            let frac_len = frac_part.len() as u32;
            let numerator = int_value * 10_i64.pow(frac_len) + frac_value;
            let denominator = 10_i64.pow(frac_len);
            Token::RealLiteral(numerator, denominator)
        } else {
            let int_value = number.parse::<i64>().unwrap_or(0);
            Token::IntLiteral(int_value)
        }
    }

    fn read_identifier(&mut self) -> Token {
        let mut identifier = String::new();
        while let Some(&ch) = self.input.peek() {
            if ch.is_ascii_alphanumeric() || ch == '_' {
                identifier.push(ch);
                self.input.next();
            } else {
                break;
            }
        }
        match identifier.as_str() {
            "true" => Token::BoolLiteral(true),
            "false" => Token::BoolLiteral(false),
            "bool" => Token::Bool,
            "int" => Token::Int,
            "real" => Token::Real,
            "string" => Token::String,
            "class" => Token::Class,
            "predicate" => Token::Predicate,
            "new" => Token::New,
            "for" => Token::For,
            "this" => Token::This,
            "void" => Token::Void,
            "return" => Token::Return,
            "fact" => Token::Fact,
            "goal" => Token::Goal,
            "or" => Token::Or,
            _ => Token::Identifier(identifier),
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.next_token();
        if token == Token::Eof { None } else { Some(token) }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lexer_basic_tokens() {
        let input = "+ - * / ( ) { } [ ] , ; = == ! != < <= > >= ";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![
            Token::Plus,
            Token::Minus,
            Token::Asterisk,
            Token::Slash,
            Token::LParen,
            Token::RParen,
            Token::LBrace,
            Token::RBrace,
            Token::LBracket,
            Token::RBracket,
            Token::Comma,
            Token::Semicolon,
            Token::Equal,
            Token::EqualEqual,
            Token::Not,
            Token::NotEqual,
            Token::LessThan,
            Token::LessEqual,
            Token::GreaterThan,
            Token::GreaterEqual,
        ];
        for expected in expected_tokens {
            let token = lexer.next_token();
            assert_eq!(token, expected);
        }
    }

    #[test]
    fn test_lexer_identifiers_and_numbers() {
        let input = "var1 var_2 123 45.67";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Identifier("var1".to_string()), Token::Identifier("var_2".to_string()), Token::IntLiteral(123), Token::RealLiteral(4567, 100)];
        for expected in expected_tokens {
            let token = lexer.next_token();
            assert_eq!(token, expected);
        }
    }

    #[test]
    fn test_lexer_string_literals() {
        let input = r#""hello" "world""#;
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::StringLiteral("hello".to_string()), Token::StringLiteral("world".to_string())];
        for expected in expected_tokens {
            let token = lexer.next_token();
            assert_eq!(token, expected);
        }
    }

    #[test]
    fn test_lexer_keywords() {
        let input = "int real string class predicate new for this void return fact goal or";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Int, Token::Real, Token::String, Token::Class, Token::Predicate, Token::New, Token::For, Token::This, Token::Void, Token::Return, Token::Fact, Token::Goal, Token::Or];
        for expected in expected_tokens {
            let token = lexer.next_token();
            assert_eq!(token, expected);
        }
    }

    #[test]
    fn test_lexer_riddle() {
        let input = "class Person { int age; string name; }";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Class, Token::Identifier("Person".to_string()), Token::LBrace, Token::Int, Token::Identifier("age".to_string()), Token::Semicolon, Token::String, Token::Identifier("name".to_string()), Token::Semicolon, Token::RBrace];
        for expected in expected_tokens {
            let token = lexer.next_token();
            assert_eq!(token, expected);
        }
    }

    #[test]
    fn test_lexer_dot_numbers() {
        let input = ".5 . .123 0.5";
        let mut lexer = Lexer::new(input);

        // .5 -> RealLiteral(5, 10)
        assert_eq!(lexer.next_token(), Token::RealLiteral(5, 10));

        // . -> Dot
        assert_eq!(lexer.next_token(), Token::Dot);

        // .123 -> RealLiteral(123, 1000)
        assert_eq!(lexer.next_token(), Token::RealLiteral(123, 1000));

        // 0.5 -> RealLiteral(5, 10)
        assert_eq!(lexer.next_token(), Token::RealLiteral(5, 10));
    }

    #[test]
    fn test_lexer_skips_single_line_comments() {
        let input = "int x; // comment\n real y;";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Int, Token::Identifier("x".to_string()), Token::Semicolon, Token::Real, Token::Identifier("y".to_string()), Token::Semicolon];

        for expected in expected_tokens {
            assert_eq!(lexer.next_token(), expected);
        }
    }

    #[test]
    fn test_lexer_skips_multiline_comments() {
        let input = "int /* ignore\nthis */ value;";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Int, Token::Identifier("value".to_string()), Token::Semicolon];

        for expected in expected_tokens {
            assert_eq!(lexer.next_token(), expected);
        }
    }

    #[test]
    fn test_lexer_slash_still_tokenized_when_not_comment() {
        let input = "a / b";
        let mut lexer = Lexer::new(input);
        let expected_tokens = vec![Token::Identifier("a".to_string()), Token::Slash, Token::Identifier("b".to_string())];

        for expected in expected_tokens {
            assert_eq!(lexer.next_token(), expected);
        }
    }
}
